Add torch.utils.show_pickle for showing pickle contents in saved models

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35168 Add torch.utils.show_pickle for showing pickle contents in saved models**

Summary:
Sometimes when a saved model isn't working, it's nice to be able to look
at the contents of the pickle files.  Unfortunately, pickletools output
isn't particularly readable, and unpickling is often either not possible
or runs so much post-processing code that it's not possible to tell
exactly what is present in the pickled data.

This script uses a custom Unpickler to unpickle (almost) any data into
stub objects that have no dependency on torch or any other runtime types
and suppress (almost) any postprocessing code.

As a convenience, the wrapper can search through zip files, supporting
command lines like

`python -m torch.utils.show_pickle /path/to/model.pt1@*/data.pkl`

When the module is invoked as main, we also install a hack in pprint to
allow semi-resonable formatting of our stub objects.

Test Plan:
Ran it on a data.pkl, constants.pkl, and a debug pkl

Differential Revision: [D20842550](https://our.internmc.facebook.com/intern/diff/D20842550)

dreiss committed:
Add torch.utils.show_pickle for showing pickle contents in saved models

Summary:
Sometimes when a saved model isn't working, it's nice to be able to look
at the contents of the pickle files.  Unfortunately, pickletools output
isn't particularly readable, and unpickling is often either not possible
or runs so much post-processing code that it's not possible to tell
exactly what is present in the pickled data.

This script uses a custom Unpickler to unpickle (almost) any data into
stub objects that have no dependency on torch or any other runtime types
and suppress (almost) any postprocessing code.

As a convenience, the wrapper can search through zip files, supporting
command lines like

`python -m torch.utils.show_pickle /path/to/model.pt1@*/data.pkl`

When the module is invoked as main, we also install a hack in pprint to
allow semi-resonable formatting of our stub objects.

Test Plan:
Ran it on a data.pkl, constants.pkl, and a debug pkl

dreiss commented:
Updated version of #28383

suo commented on test/test_show_pickle.py:
```diff
@@ -0,0 +1,32 @@
+            self.assertRegex(output, "weight")
+
+
+if __name__ == '__main__':
```
to actually get this test to run in the CI you'll need to add it to `run_test.py`

suo reviewed (approved):
lgtm

dreiss committed:
Update on "Add torch.utils.show_pickle for showing pickle contents in saved models"


Summary:
Sometimes when a saved model isn't working, it's nice to be able to look
at the contents of the pickle files.  Unfortunately, pickletools output
isn't particularly readable, and unpickling is often either not possible
or runs so much post-processing code that it's not possible to tell
exactly what is present in the pickled data.

This script uses a custom Unpickler to unpickle (almost) any data into
stub objects that have no dependency on torch or any other runtime types
and suppress (almost) any postprocessing code.

As a convenience, the wrapper can search through zip files, supporting
command lines like

`python -m torch.utils.show_pickle /path/to/model.pt1@*/data.pkl`

When the module is invoked as main, we also install a hack in pprint to
allow semi-resonable formatting of our stub objects.

Test Plan:
Ran it on a data.pkl, constants.pkl, and a debug pkl

[ghstack-poisoned]

dreiss committed:
Update on "Add torch.utils.show_pickle for showing pickle contents in saved models"


Summary:
Sometimes when a saved model isn't working, it's nice to be able to look
at the contents of the pickle files.  Unfortunately, pickletools output
isn't particularly readable, and unpickling is often either not possible
or runs so much post-processing code that it's not possible to tell
exactly what is present in the pickled data.

This script uses a custom Unpickler to unpickle (almost) any data into
stub objects that have no dependency on torch or any other runtime types
and suppress (almost) any postprocessing code.

As a convenience, the wrapper can search through zip files, supporting
command lines like

`python -m torch.utils.show_pickle /path/to/model.pt1@*/data.pkl`

When the module is invoked as main, we also install a hack in pprint to
allow semi-resonable formatting of our stub objects.

Test Plan:
Ran it on a data.pkl, constants.pkl, and a debug pkl

Differential Revision: [D20842550](https://our.internmc.facebook.com/intern/diff/D20842550)
